File: src/main/java/io/vproxy/vfx/ui/loading/VProgressBar.java
package io.vproxy.vfx.ui.loading;

import io.vproxy.base.util.callback.Callback;
import io.vproxy.vfx.manager.internal_i18n.InternalI18n;
import io.vproxy.vfx.manager.task.TaskManager;
import io.vproxy.vfx.theme.Theme;
import io.vproxy.vfx.ui.shapes.VLine;
import io.vproxy.vfx.util.FXUtils;
import io.vproxy.vfx.util.MiscUtils;
import javafx.application.Platform;
import javafx.beans.property.DoubleProperty;
import javafx.beans.property.DoublePropertyBase;
import javafx.scene.Group;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.function.Consumer;

public class VProgressBar extends Group {
    private static final double radius = 1;
    private static final double width = radius * 2;

    private double length;
    private double progress;
    private final VLine backgroundLine = new VLine(width);
    private final VLine progressLine = new VLine(width);

    public VProgressBar() {
        getChildren().addAll(backgroundLine, progressLine);
        backgroundLine.setStartX(radius);
        backgroundLine.setStroke(Theme.current().progressBarBackgroundColor());

        progressLine.setStartX(radius);
        progressLine.setStroke(Theme.current().progressBarProgressColor());
    }

    public double getLength() {
        return length;
    }

    public void setLength(double length) {
        this.length = length;
        backgroundLine.setEndX(length - radius);
        updateProgressLine();
    }

    private final DoublePropertyBase progressProperty = new DoublePropertyBase() {
        @Override
        protected void invalidated() {
            setProgress(progressProperty.get());
        }

        @Override
        public Object getBean() {
            return VProgressBar.this;
        }

        @Override
        public String getName() {
            return "progressProperty";
        }
    };

    public DoubleProperty progressProperty() {
        return progressProperty;
    }

    public double getProgress() {
        return progress;
    }

    public void setProgress(double progress) {
        if (progress < 0) {
            progress = 0;
        } else if (progress > 1) {
            progress = 1;
        }
        this.progress = progress;
        progressProperty.set(progress);
        updateProgressLine();
    }

    private void updateProgressLine() {
        var l = length * progress;
        progressLine.setEndX(l - radius);
    }

    private List<LoadingItem> items = new ArrayList<>();
    private long interval = 0;
    private Callback<Void, LoadingFailure> cb = null;
    private volatile boolean isDone = false;
    private Consumer<LoadingItem> currentLoadingItemCallback = null;

    public void setCurrentLoadingItemCallback(Consumer<LoadingItem> currentLoadingItemCallback) {
        this.currentLoadingItemCallback = currentLoadingItemCallback;
    }

    public void setItems(List<LoadingItem> items) {
        this.items = items;
    }

    public void setInterval(long interval) {
        this.interval = interval;
    }

    public void load(Callback<Void, LoadingFailure> cb) {
        this.cb = cb;

        long total = 0;
        for (var item : items) {
            total += item.weight;
        }
        loadItem(total, 0, items.iterator(), () -> Platform.runLater(() -> {
            isDone = true;
            Platform.runLater(() -> Platform.runLater(cb::succeeded));
        }));
    }

    private void loadItem(long total, long current, Iterator<LoadingItem> ite, Runnable cb) {
        if (!ite.hasNext()) {
            cb.run();
            return;
        }
        if (isDone) {
            return;
        }
        var item = ite.next();
        FXUtils.runOnFX(() -> {
            var currentCB = currentLoadingItemCallback;
            if (currentCB != null) {
                currentCB.accept(item);
            }
        });
        TaskManager.get().execute(() -> {
            final boolean ok;
            final Throwable loadingException;
            {
                boolean _ok;
                Throwable _ex;
                try {
                    _ok = item.loadFunc.getAsBoolean();
                    _ex = null;
                } catch (Throwable t) {
                    _ex = t;
                    _ok = false;
                }
                ok = _ok;
                loadingException = _ex;
            }
            if (ok) {
                if (interval > 0) {
                    MiscUtils.threadSleep(interval);
                }
            }
            Platform.runLater(() -> {
                if (!ok) {
                    isDone = true;
                    Platform.runLater(() -> this.cb.failed(new LoadingFailure(item, loadingException)));
                    return;
                }
                long newCurr = current + item.weight;
                setProgress(newCurr / (double) total);
                loadItem(total, newCurr, ite, cb);
            });
        });
    }

    public boolean terminate() {
        if (isDone) {
            return false;
        }
        isDone = true;
        FXUtils.runOnFX(() -> cb.failed(new LoadingFailure(InternalI18n.get().loadingCanceled())));
        return true;
    }
}

File: src-ts/src/ui/loading/VProgressBar.ts
import { Group } from '../../javafx/Pane.js';
import { DoubleProperty } from '../../javafx/Property.js';
import { VLine } from '../shapes/VLine.js';
import { Theme } from '../../theme/Theme.js';
import { TaskManager } from '../../manager/task/TaskManager.js';
import { FXUtils } from '../../util/FXUtils.js';
import { MiscUtils } from '../../util/MiscUtils.js';
import { Callback } from '../../vproxy-base/Callback.js';
import { InternalI18n } from '../../manager/internal_i18n/InternalI18n.js';
import { LoadingItem } from './LoadingItem.js';
import { LoadingFailure } from './LoadingFailure.js';

export class VProgressBar extends Group {
  private static readonly radius = 1;
  private static readonly width = VProgressBar.radius * 2;

  private length = 0;
  private progress = 0;
  private readonly backgroundLine: VLine = new VLine(VProgressBar.width);
  private readonly progressLine: VLine = new VLine(VProgressBar.width);

  private readonly progressProperty: DoubleProperty = new DoubleProperty(0);

  private items: LoadingItem[] = [];
  private interval = 0;
  private cb: Callback<void, LoadingFailure> | null = null;
  private isDone = false;
  private currentLoadingItemCallback: ((item: LoadingItem) => void) | null = null;

  constructor() {
    super();
    this.getChildren().addAll(this.backgroundLine, this.progressLine);
    this.backgroundLine.setStartX(VProgressBar.radius);
    this.backgroundLine.setStroke(Theme.current().progressBarBackgroundColor());

    this.progressLine.setStartX(VProgressBar.radius);
    this.progressLine.setStroke(Theme.current().progressBarProgressColor());

    this.progressProperty.addListener((_old, now) => {
      this.setProgress(now);
    });
  }

  getLength(): number {
    return this.length;
  }

  setLength(length: number): void {
    this.length = length;
    this.backgroundLine.setEndX(length - VProgressBar.radius);
    this.updateProgressLine();
  }

  progressPropertyProperty(): DoubleProperty {
    return this.progressProperty;
  }

  getProgress(): number {
    return this.progress;
  }

  setProgress(progress: number): void {
    if (progress < 0) {
      progress = 0;
    } else if (progress > 1) {
      progress = 1;
    }
    this.progress = progress;
    // Re-entry: this will fire the listener, which calls setProgress(now).
    // Since the clamped value equals what the listener received, Object.is
    // short-circuits and the call stabilizes.
    this.progressProperty.set(progress);
    this.updateProgressLine();
  }

  private updateProgressLine(): void {
    const l = this.length * this.progress;
    this.progressLine.setEndX(l - VProgressBar.radius);
  }

  setCurrentLoadingItemCallback(cb: (item: LoadingItem) => void): void {
    this.currentLoadingItemCallback = cb;
  }

  setItems(items: LoadingItem[]): void {
    this.items = items;
  }

  setInterval(interval: number): void {
    this.interval = interval;
  }

  load(cb: Callback<void, LoadingFailure>): void {
    this.cb = cb;

    let total = 0;
    for (const item of this.items) {
      total += item.weight;
    }
    this.loadItem(total, 0, 0, () => {
      FXUtils.runOnFX(() => {
        this.isDone = true;
        FXUtils.runOnFX(() => {
          FXUtils.runOnFX(() => cb.succeeded());
        });
      });
    });
  }

  private loadItem(
    total: number,
    current: number,
    index: number,
    finalCb: () => void,
  ): void {
    if (index >= this.items.length) {
      finalCb();
      return;
    }
    if (this.isDone) {
      return;
    }
    const item = this.items[index]!;
    FXUtils.runOnFX(() => {
      const currentCB = this.currentLoadingItemCallback;
      if (currentCB !== null) {
        currentCB(item);
      }
    });
    TaskManager.get().execute(async () => {
      let ok: boolean;
      let loadingException: unknown;
      try {
        ok = item.loadFunc();
        loadingException = null as unknown;
      } catch (e) {
        loadingException = e;
        ok = false;
      }
      if (ok) {
        if (this.interval > 0) {
          await MiscUtils.threadSleep(this.interval);
        }
      }
      FXUtils.runOnFX(() => {
        if (!ok) {
          this.isDone = true;
          const cb = this.cb!;
          const ex = loadingException;
          FXUtils.runOnFX(() => {
            cb.failed(new LoadingFailure(item, ex));
          });
          return;
        }
        const newCurr = current + item.weight;
        this.setProgress(newCurr / total);
        this.loadItem(total, newCurr, index + 1, finalCb);
      });
    });
  }

  terminate(): boolean {
    if (this.isDone) {
      return false;
    }
    this.isDone = true;
    const cb = this.cb!;
    FXUtils.runOnFX(() => {
      cb.failed(new LoadingFailure(InternalI18n.get().loadingCanceled()));
    });
    return true;
  }
}
